### ISSUE TITLE
Flow showing "incompatibe type" error on Promise-related functions

### ISSUE THREAD

**Issue by karelbilek** [facebook/flow#2212]
This is a very heavy simplification of my code

``` javascript
/* @flow */
function runA<X>(fn: () => (X | Promise<X>)): Promise<X>{
    return runB(fn);
}

function runB<X>(fn: () => (X | Promise<X>)): Promise<X>{
    return Promise.resolve().then(() => fn());
}
```

This shows the following (not very helpful) error:

> Promise
> This type is incompatible with some incompatible instantiation of X.

Rewriting to async doesn't help.

``` javascript
async function runA<X>(fn: () => (X | Promise<X>)): Promise<X> { // error on this line
    return runB(fn);
}

async function runB<X>(fn: () => (X | Promise<X>)): Promise<X>{
    return fn();
}
```

Reverting to older Flow fixes the error, so it's a regression.

Our library works a lot with Promises, so this is definitely annoying.

**Comment by runn1ng** [comment]
Oh. For some reason, switching the order in the function signature helps. Flow doesn't throw error here.

```javascript
function runA<X>(fn: () => (Promise<X> | X)): Promise<X> {
    return runB(fn);
}

function runB<X>(fn: () => (Promise<X> | X)): Promise<X> {
    return Promise.resolve().then(() => fn());
}
```

I thought that `A | B` and `B | A` is equivalent.

**Comment by avikchaudhuri** [comment]
@runn1ng In general, no, `A | B` is not equivalent to `B | A` when `A` and `B` may overlap (on the other hand, if they are "disjoint"...i.e., no type can be both `A` as well as `B`, then order shouldn't matter).

Basically we want ordering precisely to resolve cases of overlap. In your example, `X` is more general that `Promise<X>`, so having it appear before makes the second case never fire. I'm guessing what's happening is that you pass in a `Promise<T>`, it gets unified with the `X` case, so the return type becomes something like `Promise<Promise<T>>`...which is precisely what putting `Promise<X>` first in the order is designed to prevent. See the core Promise lib that consistently follows this ordering: https://github.com/facebook/flow/blob/master/lib/core.js#L500

**Comment by runn1ng** [comment]
....I was never listening enough during my Prolog classes, so I am bad at unification logic :) 

I still don't get what is happening there. 

So not sure if it's a bug, or not.

**Comment by avikchaudhuri** [comment]
@runn1ng Sorry, I should have explained more clearly. Let me try again.

No, this is not a bug...union cases are supposed to tried in order, with the understanding that the first case wins. So the union type `X | Promise<X>` is essentially the same as `X`, because whatever you throw at the union, the chances of it matching the first case `X` is always greater than it matching the second case `Promise<X>`, so the second case is never reached. 

So suppose you define a function whose input is `() => (X | Promise<X>)` and output type is `Promise<X>`. Let's consider some calls to that function.

Example (1): You pass something of type `() => string`. Then `string` matches `X`. Once that happens, the return type expands to `Promise<string>`.

Example (2): You pass something of type `() => Promise<number>`. Then `Promise<number>` also matches `X` (the second case is never reached)! Once that happens, the return type expands to `Promise<Promise<number>>`.

Something like example (2) is happening the code above. Changing the order of cases in the union fixes the problem, because `Promise<number>` matches `Promise<X>`, so `X` becomes `number`, and the return type correctly expands to `Promise<number>`.

**Comment by runn1ng** [comment]
Thanks! That makes sense.

**Comment by STRML** [comment]
From my understanding of TS, it chooses the more-specific case first. Am I wrong on this? If Flow is aiming to hit better TS compatibility some day, it would be nice (and less surprising) to hit this behavior.

[TS example](https://www.typescriptlang.org/play/index.html#src=class%20Promise%3CT%3E%20%7B%0D%0A%09value%3A%20T%3B%0D%0A%09constructor(value%3A%20T)%20%7B%0D%0A%09%09this.value%20%3D%20value%3B%0D%0A%09%7D%0D%0A%7D%0D%0A%0D%0Afunction%20then%3CT%3E(onFulfilled%3A%20(value%3A%20any)%20%3D%3E%20T%20%7C%20Promise%3CT%3E)%3A%20Promise%3CT%3E%3B%0D%0Afunction%20then(fn)%20%7B%0D%0A%09const%20result%20%3D%20fn('a')%3B%0D%0A%09if%20(result%20instanceof%20Promise)%20return%20result%3B%0D%0A%09return%20new%20Promise(result)%3B%0D%0A%7D%0D%0A%0D%0Avar%20a%3A%20Promise%3Cstring%3E%20%3D%20then(()%20%3D%3E%20new%20Promise('string'))%3B%0D%0Avar%20b%3A%20Promise%3Cstring%3E%20%3D%20then(()%20%3D%3E%20'string')%3B%0D%0A%0D%0A)

**Comment by karelbilek** [comment]
(I'm surprised - positively - that you keep closing 5 year old issues. Thanks!)
